src/doc/unstable-book/src/compiler-flags/source-based-code-coverage.md

```diff
@@ -156,14 +168,77 @@ There are four statistics tracked in a coverage summary:
 
 Of these four statistics, function coverage is usually the least granular while region coverage is the most granular. The project-wide totals for each statistic are listed in the summary.
 
+## Test coverage
+
+A typical use case for coverage analysis is test coverage. Rust's source-based coverage tools can both measure your tests' code coverage as percentage, and pinpoint functions and branches not tested.
+
+The following example (using the [`json5format`] crate, for demonstration purposes) show how to generate and analyze coverage results for all tests in a crate.
+
+Since `cargo test` both builds and runs the tests, we set both the additional `RUSTFLAGS`, to add the `-Zinstrument-coverage` flag, and `LLVM_PROFILE_FILE`, to set a custom filename for the raw profiling data generated during the test runs. Since there may be more than one test binary, apply `%m` in the filename pattern. This generates unique names for each test binary. (Otherwise, each executed test binary would overwrite the coverage results from the previous binary.)
+
+```shell
+$ RUSTFLAGS="-Zinstrument-coverage" \
+    LLVM_PROFILE_FILE="json5format-%m.profraw" \
+    cargo test --tests
+```
+
+Make note of the test binary file paths, displayed after the word "`Running`" in the test output:
+
+```text
+   ...
+   Compiling json5format v0.1.3 ($HOME/json5format)
+    Finished test [unoptimized + debuginfo] target(s) in 14.60s
+
+     Running target/debug/deps/json5format-fececd4653271682
+running 25 tests
+...
+test result: ok. 25 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
+
+     Running target/debug/deps/lib-30768f9c53506dc5
+running 31 tests
+...
+test result: ok. 31 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
+```
+
+You should have one ore more `.profraw` files now, one for each test binary. Run the `profdata` tool to merge them:
+
+```shell
+$ cargo profdata -- merge \
+    -sparse json5format-*.profraw -o json5format.profdata
+```
+
+Then run the `cov` tool, with the `profdata` file and all test binaries:
+
+```shell
+$ cargo cov -- report \
+    --use-color --ignore-filename-regex='/.cargo/registry' \
+    --instr-profile=json5format.profdata \
+    target/debug/deps/lib-30768f9c53506dc5 \
+    target/debug/deps/json5format-fececd4653271682
```

Comment:
my experience in https://github.com/ZcashFoundation/zebra/pull/1293 indicates that this will not produce the proper coverage report. I didn't quite get to the bottom of exactly what was happening but when I tried specifying the bins as a list of space separated arguments it would lose some of the coverage reports, and ordering mattered. My guess is it was ignoring everything after the first, or overwriting them somehow. In order to fix it I had to prefix _every_ bin with `-object`.

https://github.com/ZcashFoundation/zebra/pull/1293/commits/f4b1cd2bd0ec8958fc1e229fe5431381a1fc5852 This is the PR where I did it, the fix being adding the `printf` when we output the file names.

It might also be a good idea to suggest how to extract the list of binaries that were ran by `cargo-test`, or link to examples that do so.

```bash
cargo test --no-run --message-format=json \
    | jq -r "select(.profile.test == true) | .filenames[]" \
    | grep -v dSYM -
```

Reply:
Cool! Thanks for the recommendation. I've been struggling to get this PR to land so I may apply these suggestions in a follow-up PR.

Reply:
I've addressed this in #79818.

Thanks again!